### PR TITLE
st: 0.6 -> 0.7

### PR DESCRIPTION
**Pull request by andsild**
Note: first PR, sorry for any mistakes that might have been made
###### Motivation for this change

updating old software
###### Things done

Adding patch for delete key, from old bug; http://lists.suckless.org/dev/1212/13800.html
(the order of + and - operators are wrong)
Bump version from 0.6 -> 0.7
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

@viric was the old maintainer

### CODE REVIEW

File: pkgs/applications/misc/st/default.nix [hunk 1]
```diff
@@ -4,22 +4,20 @@
...
   
   src = fetchurl {
-    url = "http://dl.suckless.org/st/${name}.tar.gz";
-    sha256 = "0avsfc1qp8zvshsfjwwrkvk411jlqy58z225bsdhjkl1qc40qcc5";
+    url = "http://dl.suckless.org/st/${name}-${version}.tar.gz";
+    sha1 = "358bca5bfc6ee85352dcd4f45f13f78474e5de42";
   };
-
-  inherit patches;
+  patches = [ ./st.patch ];
```
Comment (pSub) [inline pkgs/applications/misc/st/default.nix]: What does this patch do?

Comment (joachifm) [inline pkgs/applications/misc/st/default.nix]: Looks like a potential accidental inclusion? At least I fail to see how swapping those values would affect the build.

Comment (ehmry) [inline pkgs/applications/misc/st/default.nix]: Seems to fix the delete key, I never noticed that it didn't work in 0.6 but it works for me now.

Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: It makes the delete key work. At least on my computers it didn't work out of the box. The patch changes the ordering of + and - operators, I think this is a bug, somewhat related to http://lists.suckless.org/dev/1212/13800.html

Should I document this in the code somehow?

Comment (ehmry) [inline pkgs/applications/misc/st/default.nix]: This looks wrong, patches is an argument to the st function, either that argument should be removed, or the argument should be appended with st.patch.
```
 { stdenv, fetchurl, pkgconfig, writeText, libX11, ncurses, libXext, libXft, fontconfig
 , conf ? null, patches ? []}:
```

**Comment by ehmry** [comment]
I would like to see the patch have a more descriptive name and append to patches, not replacing it.

File: pkgs/applications/misc/st/default.nix [hunk 2]
```diff
@@ -4,22 +4,20 @@
...
 stdenv.mkDerivation rec {
-  name = "st-0.6";
+  name = "st";
+  version = "0.7";
   
   src = fetchurl {
-    url = "http://dl.suckless.org/st/${name}.tar.gz";
-    sha256 = "0avsfc1qp8zvshsfjwwrkvk411jlqy58z225bsdhjkl1qc40qcc5";
+    url = "http://dl.suckless.org/st/${name}-${version}.tar.gz";
+    sha1 = "358bca5bfc6ee85352dcd4f45f13f78474e5de42";
```
Comment (rycee) [inline pkgs/applications/misc/st/default.nix]: Should prefer sha256 over sha1.

File: pkgs/applications/misc/st/default.nix [hunk 3]
```diff
@@ -4,22 +4,20 @@
 with stdenv.lib;
 
 stdenv.mkDerivation rec {
-  name = "st-0.6";
+  name = "st";
```
Comment (rycee) [inline pkgs/applications/misc/st/default.nix]: It is important that the version is part of the name. E.g, it should be `name = "st-0.7";` for this updated version. See the [Package naming](https://nixos.org/nixpkgs/manual/#sec-package-naming) section of the Nixpkgs manual.

File: pkgs/applications/misc/st/default.nix (hunk 2 above)
Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: The [source website](http://dl.suckless.org/st/) only provides sha1 and md5 from their directories, should I compute still compute my own sha256 sum and provide that?

[And md5's are getting deprecated, right?](https://github.com/NixOS/nix/issues/802)

File: pkgs/applications/misc/st/default.nix (hunk 3 above)
Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: thanks, will fix later

File: pkgs/applications/misc/st/default.nix (hunk 2 above)
Comment (rycee) [inline pkgs/applications/misc/st/default.nix]: Ah, sorry I missed that upstream provides sha1 sums. Then I suppose it is fine to reuse that here. In general though sha256 is preferred within Nixpkgs.

Personally I usually wouldn't put the upstream hash code but instead verify the upstream hash or GPG signature and then calculate a sha256 hash to put into Nixpkgs.

Comment (aneeshusa) [inline pkgs/applications/misc/st/default.nix]: Please don't downgrade to sha1, but instead keep using sha256. Even if upstream provides a weaker hash we can use a stronger hash and provide more protection to nixpkgs users. We're already [deprecating md5](https://github.com/NixOS/nixpkgs/issues/4491) and I hope sha1 goes soon, too!

**Comment by andsild** [comment]
@ehmry I've update the commit, if you time to re-review that would be good :)

File: pkgs/applications/misc/st/default.nix [hunk 4]
```diff
@@ -4,30 +4,28 @@
...
+  name = "st-0.7";
   
   src = fetchurl {
     url = "http://dl.suckless.org/st/${name}.tar.gz";
-    sha256 = "0avsfc1qp8zvshsfjwwrkvk411jlqy58z225bsdhjkl1qc40qcc5";
+    sha256 = "f7870d906ccc988926eef2cc98950a99cc78725b685e934c422c03c1234e6000";
   };
-
-  inherit patches;
+
```
Comment (shlevy) [inline pkgs/applications/misc/st/default.nix]: Please fix leading whitespace

File: pkgs/applications/misc/st/default.nix [hunk 5]
```diff
@@ -4,30 +4,28 @@
...
   
   src = fetchurl {
     url = "http://dl.suckless.org/st/${name}.tar.gz";
-    sha256 = "0avsfc1qp8zvshsfjwwrkvk411jlqy58z225bsdhjkl1qc40qcc5";
+    sha256 = "f7870d906ccc988926eef2cc98950a99cc78725b685e934c422c03c1234e6000";
   };
-
-  inherit patches;
+  
+  patches = with self; [ ./st-fix-deletekey.patch ];
```
Comment (shlevy) [inline pkgs/applications/misc/st/default.nix]: This still does not apply the old patches. The `with self;` only brings the attributes of`self` into scope, in fact there *is* no `self` in scope but this doesn't fail because the list doesn't reference any variables that might belong to self. You probably want to do something like `let patches' = patches; in` outside of the set and then have `patches = patches' ++ [ ./st-fix-deletekey.patch ]` here

File: pkgs/applications/misc/st/default.nix [hunk 6]
```diff
@@ -4,30 +4,28 @@
...
-
   installPhase = ''
     TERMINFO=$out/share/terminfo make install PREFIX=$out
   '';
     
   meta = {
     homepage = http://st.suckless.org/;
     license = stdenv.lib.licenses.mit;
-    maintainers = with maintainers; [viric];
+    maintainers = with maintainers; [andsild];
```
Comment (shlevy) [inline pkgs/applications/misc/st/default.nix]: Why did you remove @viric? There can be multiple maintainers, have they told you they're no longer maintaining this?

Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: No, I didn't think that one through well enough, I'll put him back in

**Comment by grahamc** [comment]
FWIW you're doing a good job here :)

File: pkgs/applications/misc/st/default.nix [hunk 7]
```diff
@@ -1,33 +1,32 @@
...
-, conf ? null, patches ? []}:
+{ stdenv, fetchurl, pkgconfig, writeText, libX11, ncurses, libXext, libXft
+, fontconfig, conf ? null, patches ? []}:
 
 with stdenv.lib;
 
-stdenv.mkDerivation rec {
-  name = "st-0.6";
-  
+let patches' = if isNull patches then [] else patches;
```
Comment (joachifm) [inline pkgs/applications/misc/st/default.nix]: What is the motivation for this construct?

Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: Patches can be given by user.
We want to also apply our own patch in the code down below (st-fix-deletekey) ([given here](https://github.com/andsild/nixpkgs/blob/29aac270898118a446255ec00540e5ef258f802c/pkgs/applications/misc/st/default.nix#L15))

[Here](https://github.com/andsild/nixpkgs/blob/2e826713a3b94812afd90d4cc8a8cb004dfc186d/pkgs/top-level/all-packages.nix#L14740) however, "patches" is given default null, so we must add the if statement to make it non-null.

The ideal would be to just do `patches = patches ++ [ ./st-fix-deletekey.patch ]` but since we are in a recursive construct, that gives infinite recursion. Therefore the "let" construct outside the rec block

Comment (shlevy) [inline pkgs/applications/misc/st/default.nix]: 1. all-packages.nix passes in `patches` as `null` if the relevant config value is not set
2. `mkDerivation` is called with a recursive attribute set, so `patches` is bound to the `patches` attribute, not the `patches` argument, inside the set.

Comment (joachifm) [inline pkgs/applications/misc/st/default.nix]: The recursive issue is obvious, but having two ways of saying "no additional patches" seems odd to me. Is there a benefit to passing `null` over simply `[]`?

Comment (andsild) [inline pkgs/applications/misc/st/default.nix]: not that I am aware of, actually. I was caught up editing default.nix so it didn't occur to me

Comment (joachifm) [inline pkgs/applications/misc/st/default.nix]: Doesn't matter, just looks a bit odd (to me).
